### PR TITLE
feat(memory): AI-aware Stage-1 (path B) + source trust-tier

### PR DESCRIPTION
**Pull request by wehos**
## Summary

- 引入 **path B**：piggyback 在现有 SignalLoop (path A) 循环里、每 3 次 A tick 触发一次的 AI-aware Stage-1 only 通路。**抢回 PR #1346 之后失明的 entity=neko / proactive 屏幕上下文 grounded fact**
- fact schema 加 `source` ∈ {`'user_observation'`, `'ai_disclosure'`}；trust-tier 隔离：ai_disclosure fact **不进 Stage-2 evidence loop**（防 AI 自我披露触发 reflection reinforce 形成自我强化死循环）
- **Monotonic source upgrade**：user 印证后 ai_disclosure fact 可升级到 user_observation 并重入 Stage-2；反向永不降级
- 15 个新 regression test + 79 个 adjacent suite test 全过

## Root cause

PR #1346 (5-13) 把 ON-mode 的 per-turn fact extraction 整段砍掉（"per-turn 太贵，全交 batch"），但 SignalLoop 从 PR #929 起就只看 user msg（`_extract_user_messages_from_rows` 硬编码 `type=='human'`）。结果 ON-mode 唯一活路径完全失明 AI msg → entity=neko 类 fact 永不入库。

实测今日 facts.json：5 条 today's facts **全是 entity=master**（"博士 likes salmon / prefers sashimi / loves writing code..."），零条关于悠怡自己 / 零条 grounded 在 AI 引入的屏幕上下文。

## 设计要点（thread 拍板）

| Dimension | 选择 | 理由 |
|---|---|---|
| **D1** Sequencing | A→B | 避免 source 字段方向错（B 先写 ai_disclosure，A 后撞会丢 user_observation 升级语义） |
| **D2** B input | role-tagged 全消息 + prompt 塞 A 已抽 fact 作"已知池" | 让 LLM 输出层主动去重 path A 已抽内容 |
| **D3** source 处理 | A wins 单调升级 | user 印证不可逆 |
| **D4** B cadence | piggyback 在 A 循环里，每 N=3 次 A tick 触发；窗口下游边界用 `last_a_msg_ts`（A 实际处理过的最晚 msg）避免 race | tempo 跟着对话强度；不需要独立 wall-clock cadence；cold-start lookback 从 N×IDLE_MIN 推算 |
| **D5** Stage-2 input | `source != 'ai_disclosure'` filter，AI fact 写盘时 `signal_processed=True` 双重防御 | 干净隔离自我强化死循环 |
| **D6** reflection synthesis | 不 filter | ai_disclosure 可入合成池，但 evidence 涨分仍只靠 user_observation（confirmed 仍需 user 印证） |

## 成本

| | A 单 call | B 单 call | 日 cost |
|---|---|---|---|
| input tok | ~4300 | ~3700-7300 | A: 12×4300=52k；B: 2-4×~5500=16k；总 ~68k/天 (+~30% vs A-only) |
| LLM calls | 24 | 2-4 | 总 26-28 vs A-only 24 |

月成本（summary tier 中端模型）：~$1.5-4/月 vs A-only $1-3/月，**净增量 < $1/月** 单角色。

## Test plan

- [x] 新增 `tests/unit/test_ai_aware_stage1_path_b.py` (15 cases): role-tagged 提取 / source 持久化 / LLM 显式 source 优先 / monotonic upgrade（ai→user 升级 + 重置 signal_processed；user→ai 永不降级）/ Stage-2 unprocessed pool 排除 ai_disclosure / cold-start lookback 从 N×IDLE_MIN 推算 / A 未跑过时 B 跳过 / B cursor 单调推进 / 已知池过滤+cap+排序 / 源码扫描 wiring 不变量
- [x] `uv run pytest tests/unit/test_ai_aware_stage1_path_b.py -q` → 15 passed
- [x] adjacent suite (`evidence_extraction / memory_server_cache_settle / reflection_synthesis_loop / memory_recall / signal_loop_message_wire / reflection_idempotent / reflection_filters`) 79 passed, 无回归
- [ ] **本地手动验证**：跑 launcher、跟猫娘聊几轮含 AI 自我披露的对话、看 `[PathB] {name}: AI-aware Stage-1 抽出 N 条新 fact` 日志、看 facts.json 出现 `source='ai_disclosure'` 的 entity=neko fact

## Follow-up issues（不在本 PR scope）

- FactDedupResolver merge/replace 应该 archive-style 而不是物理删除（pre-existing inconsistency，dedup 命中后原文从 hybrid_recall 池消失。本 PR 不动）
- review_history 路径耦合：曾考虑作为 AI-aware 触发点，最终因 race / compress 边界 / 重复抽取风险被否决

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **新功能**
  * 增加 AI‑感知的补抓流程（Path B），按角色抽取并裁剪为首尾用户发言区间；新增本地化 AI‑感知事实抽取提示；事实记录新增来源层级（user_observation / ai_disclosure）。

* **改进**
  * 可配置触发节奏、窗口消息上限与已知 fact 池上限；来源单向升级与重试保留语义；时区归一为 naive；微秒级游标推进避免死循环；AI 自我披露被排除于证据循环；仅在有用户发言时触发 Path B。

* **测试**
  * 新增大量单元测试覆盖消息抽取、窗口裁剪、来源持久化、游标/时区/微秒边界与触发节奏。

[Review Change Stack](https://app.coderabbit.ai/change-stack/Project-N-E-K-O/N.E.K.O/pull/1408?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: app/memory_server.py [hunk 1]
```diff
@@ -1760,6 +1812,113 @@ async def _adispatch_evidence_signals(
...
+    if persisted:
+        logger.info(
+            f"[PathB] {name}: AI-aware Stage-1 抽出 {len(persisted)} 条新 fact "
+            f"(window={last_b.isoformat(timespec='seconds')} → "
+            f"{last_a_msg_ts.isoformat(timespec='seconds')}, "
+            f"known_pool={len(known_pool)})"
+        )
+
+    # 成功跑完（无论抽出 0 还是 N 条），推进 B cursor 到 A 处理过的最晚点
+    state['last_b_check_ts'] = last_a_msg_ts
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py]: 抓得对。修了 (commit `5cf6f9b9f`): cursor 推到 SQL 实际取到的最后一行 ts (`_coerce_db_ts(rows[-1][0])`) 而不是 `last_a_msg_ts`。无截断时 last fetched == last_a_msg_ts 语义不变；有截断时 < last_a_msg_ts，下次 B trigger 从该点继续覆盖剩余尾巴。加 regression test 钉死 truncation 场景下 cursor 必须 < last_a_msg_ts。

File: app/memory_server.py [hunk 2]
```diff
@@ -1760,6 +1812,113 @@ async def _adispatch_evidence_signals(
...
+        created_at_raw = f.get('created_at') or ''
+        try:
+            created_at = datetime.fromisoformat(created_at_raw[:19])
+        except (ValueError, TypeError):
+            continue
+        if last_b <= created_at <= last_a_msg_ts:
+            known_pool.append(f)
+    known_pool.sort(
+        key=lambda f: -int(f.get('importance', 5) or 5),
+    )
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py]: 同样真 bug。改用现成的 `safe_importance` helper (`memory/facts.py:61` 已存在的兜底，专门处理 schema 漂移)。加 regression test：5 条 dirty importance ("high" / None / list / 缺字段) 不抛、全进 pool。Commit `5cf6f9b9f`。

File: app/memory_server.py [hunk 3]
```diff
@@ -1760,6 +1812,113 @@ async def _adispatch_evidence_signals(
...
+    known_pool = known_pool[:MAX_KNOWN_POOL_FACTS]
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py]: 抓得对，已修 (commit `3fa5768a6`)。去掉上界 `created_at <= last_a_msg_ts`，只留下界 `created_at >= last_b`：A 的 idle/polling 延迟让最新一批 A facts 的 created_at 普遍 > last_a_msg_ts，设上界把它们整批漏掉是真的让 known_pool 对最新 A 抽取失效。新加 regression test `test_path_b_known_pool_includes_just_written_a_facts_despite_clock_skew` 钉死该契约。不走"caller 传 A batch"方案：B 是 piggyback 在 A 循环但路径解耦，不想给 A 加 callback 把内部 batch 漏出来。

File: memory/facts.py [hunk 4]
```diff
@@ -987,6 +1061,82 @@ async def extract_facts(self, messages: list, lanlan_name: str) -> list[dict]:
...
+        后续 A tick（A 自己的 Stage-1 失败有独立的 FactExtractionFailed 处理）。
+        """
+        extracted = await self._allm_extract_facts_with_known_pool(
+            lanlan_name, messages, known_pool,
+        )
+        if not extracted:
+            return []
+        return await self._apersist_new_facts(
+            lanlan_name, extracted, default_source='ai_disclosure',
+        )
```
[the remark this answers is not in the document]
Comment (wehos) [reply memory/facts.py:1109]: 真 bug，已修 (commit `3fa5768a6`)。`aextract_facts_with_known_pool` 区分 None (Stage-1 终态失败) vs [] (LLM 成功 0 抽)，`_run_path_b` 检测到 None 保留 cursor 下次重试。Codex 同一 commit 也提了完全相同的 P1，一并按 None 透传方案修了。

File: tests/unit/test_ai_aware_stage1_path_b.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+    }
+
+    fake_time_manager = MagicMock()
+    # 窗口完全空
+    fake_time_manager.aretrieve_original_by_timeframe = AsyncMock(return_value=[])
+
+    with patch.object(memory_server, 'time_manager', fake_time_manager):
+        await memory_server._run_path_b('悠怡', state)
+
+    assert state['last_b_check_ts'] == last_a_msg_ts
```
[the remark this answers is not in the document]
Comment (wehos) [reply tests/unit/test_ai_aware_stage1_path_b.py]: 补了 `test_path_b_stage1_terminal_failure_preserves_cursor` (commit `3fa5768a6`)：mock `aextract_facts_with_known_pool` 返 None，断言 `state[last_b_check_ts]` 保持原值不推进。

File: memory/facts.py [hunk 6]
```diff
@@ -987,6 +1061,82 @@ async def extract_facts(self, messages: list, lanlan_name: str) -> list[dict]:
...
+        - Stage-2 不走（path B 设计就不进 evidence loop）
+
+        Stage-1 失败时返 [] 而非 raise——path B 是补抓性质，失败不该阻塞
+        后续 A tick（A 自己的 Stage-1 失败有独立的 FactExtractionFailed 处理）。
+        """
+        extracted = await self._allm_extract_facts_with_known_pool(
+            lanlan_name, messages, known_pool,
+        )
+        if not extracted:
+            return []
```
[the remark this answers is not in the document]
Comment (wehos) [reply memory/facts.py:1106]: 已修 (commit `3fa5768a6`)。`aextract_facts_with_known_pool` 改返 `list | None`：None 表 Stage-1 LLM 终态失败，[] 表成功但 0 抽。`_run_path_b` 检 None 后 `logger.warning + return`，cursor 不动下次同窗口重试。补了 regression test `test_path_b_stage1_terminal_failure_preserves_cursor`。

File: app/memory_server.py [hunk 7]
```diff
@@ -1760,6 +1812,126 @@ async def _adispatch_evidence_signals(
...
+        return
+    if not rows:
+        # 窗口内无 msg（A 处理过但都是空白？少见但可能）。推 cursor 跳过。
+        state['last_b_check_ts'] = last_a_msg_ts
+        return
+
+    message_dicts = _extract_role_tagged_messages_from_rows(rows)
+    if not message_dicts:
+        # 全是 system msg / 空内容。推 cursor 跳过。
+        state['last_b_check_ts'] = last_a_msg_ts
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py]: 真 bug，已修 (commit `3fa5768a6`)。把 `last_fetched_ts = _coerce_db_ts(rows[-1][0])` 抽到 `_extract_role_tagged_messages_from_rows` filter 之前算，所有 cursor 推进点统一用 last_fetched_ts（不再用 last_a_msg_ts）。无截断时 last_fetched == last_a_msg_ts 语义不变，截断时严格 <。新 regression test `test_path_b_truncated_window_all_filtered_advances_to_last_fetched`：2 行 system msg + 第 2 行 ts < last_a_msg_ts，断言 cursor 推到截断点。

File: tests/unit/test_ai_aware_stage1_path_b.py [hunk 8]
```diff
@@ -0,0 +1,645 @@
...
+    fs._apersist_new_facts = AsyncMock(return_value=[])
+    fs.aload_facts = AsyncMock(return_value=fake_facts)
+    fs._aload_signal_targets = AsyncMock(return_value=[
+        {'id': 'obs1', 'text': '观察', 'target_type': 'reflection'},
+    ])
+    fs._allm_detect_signals = AsyncMock(return_value=[])  # signals=[] but ran
+
+    # _allm_detect_signals 被调用时传的 batch 就是过 filter 后的 unprocessed。
+    # 三元 return 故意全用 _ 前缀——测试关心的是 mock call 参数，不是返回值。
+    _persisted, _signals, _batch_ids = await fs.aextract_facts_and_detect_signals(
```
[the remark this answers is not in the document]
Comment (wehos) [reply tests/unit/test_ai_aware_stage1_path_b.py]: 已修 (commit `3fa5768a6`)：三元 return 完全不绑变量，直接 `await fs.aextract_facts_and_detect_signals(...)`，CodeQL 不再判 unused（连 `_xxx` 前缀也被它判 unused，故彻底不绑）。

[the remark this answers is not in the document]
Comment (wehos) [reply tests/unit/test_ai_aware_stage1_path_b.py]: 同上 (commit `3fa5768a6`)。

[the remark this answers is not in the document]
Comment (wehos) [reply tests/unit/test_ai_aware_stage1_path_b.py]: 同上 (commit `3fa5768a6`)。

File: app/memory_server.py [hunk 9]
```diff
@@ -1890,6 +2088,34 @@ async def _signal_check_one(name: str):
...
+                if last_msg_ts is not None:
+                    state['last_a_msg_ts'] = last_msg_ts
+
+                # Path B trigger：A 成功跑完后 bump counter；达 N 触发
+                # _run_path_b（AI-aware Stage-1 only，详见函数 docstring）。
+                state['b_tick_counter'] = state.get('b_tick_counter', 0) + 1
+                if state['b_tick_counter'] >= EVIDENCE_AI_AWARE_EVERY_N_A_TICKS:
+                    state['b_tick_counter'] = 0
+                    try:
+                        await _run_path_b(name, state)
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py:2237]: 抓得对，这是真 P1——path B 的主要使命就是补抓 proactive / AI-only 窗口的 AI 自我披露，被自己 user-only 早 return 路径漏掉等于把 B 阉了。修了 (commit `c0931b1d5`)：把 `last_a_msg_ts` 更新提到 user_msgs 检查之前 (state 共享给两个分支)，AI-only 早 return 路径里也 bump b_tick_counter + maybe trigger _run_path_b。Regression test `test_path_b_triggered_for_ai_only_window_no_user_msgs` 用源码扫描钉死这两个要素。

File: app/memory_server.py [hunk 10]
```diff
@@ -1760,6 +1812,152 @@ async def _adispatch_evidence_signals(
...
+        return
+    if persisted:
+        logger.info(
+            f"[PathB] {name}: AI-aware Stage-1 抽出 {len(persisted)} 条新 fact "
+            f"(window={last_b.isoformat(timespec='seconds')} → "
+            f"{last_fetched_ts.isoformat(timespec='seconds')}, "
+            f"known_pool={len(known_pool)})"
+        )
+
+    state['last_b_check_ts'] = last_fetched_ts
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py:2076]: 真 bug，已修 (commit `c0931b1d5`)。检测：`len(rows) >= MAX_AI_AWARE_WINDOW_MSGS AND first_fetched_ts == last_fetched_ts` → cursor +1μs 越过该 ts，下次 BETWEEN 不会再捞同批。代价：该 ts 簇 LIMIT 之后的 tail row 被 skip（罕见——正常 turn 写 2~5 行远 < 200）。warning log 把 ts + 簇大小打出来便于事后排查。不上 (ts, rowid) 复合 cursor + 改写 SQL：成本太高对一个 corner case 不划算。2 条对偶 regression test：同 ts 簇 →+1μs；diverse ts 截断 → 不 epsilon。

File: app/memory_server.py [hunk 11]
```diff
@@ -1821,9 +2042,36 @@ async def _signal_check_one(name: str):
...
+                    if state['b_tick_counter'] >= EVIDENCE_AI_AWARE_EVERY_N_A_TICKS:
+                        state['b_tick_counter'] = 0
+                        try:
+                            await _run_path_b(name, state)
+                        except Exception as e:
+                            logger.warning(
+                                f"[PathB] {name}: AI-only 窗口 B 触发失败 "
+                                f"(skip 本轮): {e}"
+                            )
                     return
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py]: 抓得很准——这是真 bug，round-3 的 AI-only 分支 fix 形同虚设。修了 (commit `6b07925ac`)：去掉 `_run_post_turn_signals` 里 `if user_msgs:` 对 `_signal_check_record_turn` 的 gate，无条件 bump。代价：AI-only burst 让 idle gate 提前到、`_signal_check_one` 多跑几次，但内部对 user_msgs_text 空有早 return（round-3 加的），Stage-1 LLM 不会白白跑——多一次 SQL 读 + path B 触发判定。Regression test `test_post_turn_signals_bumps_counter_for_ai_only_turn` 用源码扫描钉死 counter bump 不能再被 user_msgs gate 挡掉。

File: app/memory_server.py [hunk 12]
```diff
@@ -1760,6 +1812,175 @@ async def _adispatch_evidence_signals(
...
+        len(rows) >= MAX_AI_AWARE_WINDOW_MSGS
+        and first_fetched_ts is not None
+        and last_fetched_ts == first_fetched_ts
+    ):
+        logger.warning(
+            f"[PathB] {name}: 同 ts 簇 {first_fetched_ts.isoformat(timespec='microseconds')} "
+            f"行数 ≥ LIMIT ({MAX_AI_AWARE_WINDOW_MSGS})，cursor +1μs 越过避免死循环；"
+            f"该 ts 簇 LIMIT 之后的 tail row 会被 skip"
+        )
+        last_fetched_ts = last_fetched_ts + timedelta(microseconds=1)
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py:1982]: 这轮不修，trade-off intentional——commit `c0931b1d5` 已经在 inline + commit msg 文档化：(ts, rowid) 复合 cursor + 改写 SQL 的 proper fix 对一个 corner case 太重，同 ts 簇 ≥ LIMIT (200) 在正常对话（一次 turn 2~5 行）下不可能命中，仅 bulk import 风险。和上游用户提的"毒窗口无限重试 liveness 兜底"打包到独立 follow-up issue 一起处理（已 spawn task）：两者都是"窗口卡死/数据丢失" corner case，单独修很零碎，应当统一 design max-retry / bypass / drain 策略后一起实现。这轮 PR 范围 hold 在 path B 设计本身。

File: app/memory_server.py [hunk 13]
```diff
@@ -1760,6 +1812,175 @@ async def _adispatch_evidence_signals(
...
+        rows = await time_manager.aretrieve_original_by_timeframe(
+            name, last_b, last_a_msg_ts,
+            limit_rows=MAX_AI_AWARE_WINDOW_MSGS,
+        )
+    except Exception as e:
+        logger.warning(f"[PathB] {name}: 读取窗口失败: {e}")
+        return
+    if not rows:
+        # 窗口内无 msg（A 处理过但都是空白？少见但可能）。推 cursor 跳过。
+        state['last_b_check_ts'] = last_a_msg_ts
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py]: 抓得对，真 bug 已修 (commit `ffc6b2c54`)。`time_manager.aretrieve_original_by_timeframe` 在 SQL exception / engine init 失败 / 维护态都 swallow + 返 []（timeindex.py:374-394），caller 端不可区分 → transient 失败下 path B 静默丢 fact。修法：rows 空时 cursor 不推，下次 B trigger 重试同窗口（真空窗口下次再 query 0 rows 也是 SQLite 空范围常数代价）。原 `test_path_b_window_empty_still_advances_cursor` 测的是旧错行为，改名为 `test_path_b_empty_rows_preserves_cursor_for_retry` + 断言反过来。path A / rebuttal loop 有同款 latent bug，但 PR 范围 hold 在 path B；和上游用户提的"毒窗口 max-retry 兜底"打包到 spawned liveness follow-up task systematic 修。

File: app/memory_server.py [hunk 14]
```diff
@@ -1760,6 +1812,187 @@ async def _adispatch_evidence_signals(
...
+    last_b = state.get('last_b_check_ts')
+    if last_b is None:
+        # Cold start：B 第一次 trigger 时已等了 EVIDENCE_AI_AWARE_EVERY_N_A_TICKS
+        # 个 A tick。每个 A tick 平均覆盖 EVIDENCE_SIGNAL_CHECK_IDLE_MINUTES 分钟
+        # （idle gate 阈值），所以 A 已覆盖范围 ≈ N × IDLE_MINUTES 分钟。
+        # 用这个推算作为 B 第一次窗口起点。
+        estimated_a_coverage = timedelta(
+            minutes=EVIDENCE_AI_AWARE_EVERY_N_A_TICKS * EVIDENCE_SIGNAL_CHECK_IDLE_MINUTES
+        )
+        last_b = last_a_msg_ts - estimated_a_coverage
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py:1919]: 抓得对，已修 (commit `520b993cf`)。cold-start lookback 改用 `max(EVIDENCE_AI_AWARE_EVERY_N_A_TICKS, EVIDENCE_SIGNAL_CHECK_EVERY_N_TURNS) × EVIDENCE_SIGNAL_CHECK_IDLE_MINUTES`，cover turn-count gate / sparse turn 两个 A 触发节律的较保守值。默认 max(3, 10)×10min = 100min，比原 30min 宽 3.3 倍。仍只从已 imported 常数推算不引新魔法。代价：cold-start 单次 SQL 扫描范围变宽，由 MAX_AI_AWARE_WINDOW_MSGS LIMIT + Stage-1 dedup hash 兜底——overshoot 安全。Regression test `test_path_b_cold_start_uses_max_of_ticks_and_turns` 用默认 N_TURNS > N_TICKS 锁死 max 语义。

File: app/memory_server.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py:2237]: 更新：我之前承认是 bug 走反了，已 revert (commit `3cc85f66a`)。重新评估后这是 product thesis 而非 bug：

- `turns_since` 不是消息总数，是 **user engagement** 信号——counter 累积 user 发声，AI 自言自语不算
- Path B 是 piggyback path A 的 trigger 跑（不独立调度），自然跟着 user-engagement 语义。User 没参与的窗口里 A 不抽 user_observation，B 也跟着不拣 AI 自我披露
- N.E.K.O. 产品命题："90% 没心没肺 + 10% 神明降临"。AI 自言自语 + user 不搭理 = 廉价层内容，**故意**不该被自动当 fact 沉淀污染 memory。只有 user 印证过的才升级到神明降临层

Path B 的 known_pool 拣回机制设计目的是：用户有参与的窗口里 SQL/LIMIT 截断时漏掉的 AI msg 拣回来——而不是把纯 proactive AI 独白全量收进 memory。

新加 `test_path_b_skipped_for_ai_only_window_no_user_msgs` 反向钉死该契约。原 fix 加的 regression test 一并 revert。

File: app/memory_server.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py]: 更新：之前承认 bug 是走反了，已 revert (commit `3cc85f66a`)。详见 thread #3258587175 我对 Codex round-3 P1 的同主题回复——简言之 `turns_since` gate 是 user engagement 信号（不是消息总数）就是设计意图，product thesis "90% 没心没肺 + 10% 神明降临" 决定 AI 自言自语不该自动进 memory。`_run_post_turn_signals` 的 `if user_msgs:` gate 恢复了，对应 regression test 反向钉死该契约。

File: app/memory_server.py [hunk 15]
```diff
@@ -1760,6 +1812,198 @@ async def _adispatch_evidence_signals(
...
+    for f in all_facts:
+        if not isinstance(f, dict):
+            continue
+        created_at_raw = f.get('created_at') or ''
+        try:
+            created_at = datetime.fromisoformat(created_at_raw[:19])
+        except (ValueError, TypeError):
+            continue
+        if created_at >= last_b:
+            known_pool.append(f)
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py:2050]: 抓得对，已修 (commit `404445bd3`)。去掉 `[:19]` 截秒，直接 `datetime.fromisoformat(created_at_raw)` 保留微秒；`created_at` 写盘用 `datetime.now().isoformat()` 含微秒，try/except 仍兜底脏数据。Regression test `test_path_b_known_pool_preserves_microsecond_precision_at_boundary`：last_b = T+0.5s、fact created_at = T+0.7s，断言进 pool；若回归 `[:19]`，fact 会被解成 T+0.0 → 误排 → test 失败。

File: app/memory_server.py [hunk 16]
```diff
@@ -1760,6 +1833,219 @@ async def _adispatch_evidence_signals(
...
+            continue
+        created_at_raw = f.get('created_at') or ''
+        try:
+            # 完整 ISO 解析（含微秒）—— `created_at` 是 datetime.now().isoformat()
+            # 写盘的，截到 [:19] 会丢微秒，让 created_at == last_b + 0.x 秒的
+            # fact 在 `>= last_b` 比较里被误判出窗口（CodeRabbit on PR #1408）。
+            created_at = datetime.fromisoformat(created_at_raw)
+        except (ValueError, TypeError):
+            continue
+        if created_at >= last_b:
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py:2049]: 抓得对，已修 (commit `782f9e7dc`)。本仓库 `_apersist_new_facts` 写的是 `datetime.now().isoformat()` (naive)，理论上不会触发；但 import/migration 路径不可控，TZ-aware 进来会让 `_run_path_b` 永久哑火。修法：parsed `created_at` 若有 tzinfo 就 `replace(tzinfo=None)`，比较口径当 naive 处理。Regression test `test_path_b_known_pool_normalizes_tz_aware_created_at` 用 `tzinfo=timezone.utc` 构 3 条 fact，断言不抛 + 全进 pool。

File: app/memory_server.py [hunk 17]
```diff
@@ -1760,6 +1833,227 @@ async def _adispatch_evidence_signals(
...
+            continue
+        # 防御：本仓库 `_apersist_new_facts` 写的 `created_at` 都是 naive
+        # datetime.now().isoformat()，但若 import/migration 路径写入了 TZ-aware
+        # 值（如 "...+00:00"），跟 naive 的 last_b 比较会抛 TypeError 让
+        # `_run_path_b` 一直 fail，path B 对该角色永久哑火（Codex P1 round-7
+        # on PR #1408）。比较口径上把 aware 当 naive 处理——绝大多数场景就是
+        # 同一 wall-clock 时间，时区差异不应让 fact 抽取整段挂掉。
+        if created_at.tzinfo is not None:
+            created_at = created_at.replace(tzinfo=None)
+        if created_at >= last_b:
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py:2049]: 对，round-7 修了 created_at 一侧但漏了 cursor 一侧，对称 bug。修法 (commit `b4ac9f1f9`) 上两道闸：

1. **Bottleneck**: `_coerce_db_ts` 解出 TZ-aware 时统一 `replace(tzinfo=None)`。所有 SQL row ts 过这个函数，cursor 流域一刀切，覆盖度最广
2. **Belt-and-suspenders**: `_run_path_b` 入口再对 state 里的 `last_a_msg_ts` / `last_b_check_ts` strip 一遍。外部 state injection 或升级前残留的 aware 仍能漏进来，这道闸把整个 `_run_path_b` 变 naive-only 域

2 条 regression test：`_coerce_db_ts` 单元 (aware datetime / aware ISO string / naive 不变) + end-to-end (state 注入 aware `last_a_msg_ts` + aware SQL row + naive facts.json，断言不抛 + 推进后 cursor 是 naive)。

File: memory/facts.py [hunk 18]
```diff
@@ -987,6 +1061,92 @@ async def extract_facts(self, messages: list, lanlan_name: str) -> list[dict]:
...
+            call_type="memory_fact_extraction_ai_aware",
+        )
+        if extracted is None:
+            return None
+        if not isinstance(extracted, list):
+            logger.warning(
+                f"[FactStore] {lanlan_name}: path-B Stage-1 返回非数组 "
+                f"{type(extracted).__name__}，当作空列表处理"
+            )
+            return []
```
[the remark this answers is not in the document]
Comment (wehos) [reply memory/facts.py]: 抓得对，跟 round-2 的"terminal None 折叠 []"是同一类语义的对偶 fix。已修 (commit `1bc12a056`)：`_allm_extract_facts_with_known_pool` 非 list payload 返 None 等同 terminal failure，让 `_run_path_b` 保留 cursor 下次重试。Docstring Returns 段同步更新，明确 None 触发条件包含"终态失败 / 错形态 payload"两种。2 条 regression test：5 种 bad payload (`{"facts": ...}` / str / int / 空对象 / bool) → None；wrapper 透传 None + persist 不调用。

File: memory/facts.py [hunk 19]
```diff
@@ -494,9 +524,27 @@ async def _apersist_new_facts(
...
+            raw_source = fact.get('source')
+            if raw_source in self._SOURCE_VALUES:
+                source = raw_source
+            else:
+                source = default_source
+
+            # Stage 1: SHA-256 exact dedup（+ source monotonic upgrade）
             content_hash = hashlib.sha256(text.encode()).hexdigest()[:16]
             if content_hash in existing_hashes:
+                existing = hash_to_existing.get(content_hash)
```
[the remark this answers is not in the document]
Comment (wehos) [reply memory/facts.py:537]: 抓得对，真 bug 已修 (commit `dc4c022a2`)。`hash_to_existing` append 时同步更新一行——之前只更 `existing_hashes` 集合，第二条同 text fact 命中 `in existing_hashes` 但 `hash_to_existing.get()` 返 None → upgrade 守卫不通过 → 静默丢弃 user_observation 升级。Regression test `test_apersist_monotonic_source_upgrade_within_same_batch` 钉死单 batch 同 text 先 ai_disclosure 后 user_observation → 落盘 1 条 + source 升 + signal_processed=False。

**Comment by wehos** [comment]
@coderabbitai 两条 nitpick 都看到了，都选择不动：

1. **AST 替代源码字符串扫描** (test_post_turn_signals_only_bumps_counter_for_user_turn / test_path_b_skipped_for_ai_only_window_no_user_msgs): 现在的 substring scan 在排版/注释扰动时确实更脆，但触发真实回归（去掉 gate / 调进 _run_path_b）时也会**响亮地失败**，pin 的契约是清晰的。AST 重写多 ~50 行机械样板，pin 的契约一样、可读性反而更差。当前 fragility 是可控 tradeoff，跳过。

2. **Monkeypatch 替代 `assert N_TURNS > N_TICKS` 默认序断言** (test_path_b_cold_start_uses_max_of_ticks_and_turns): 当前 precondition assert 是**自我文档**——告诉读者"本 test 只有在默认 N_TURNS > N_TICKS 时才有意义"。如果产品调默认值让两者退化，test 会在 precondition 处明确报错（带注释解释为什么），而不是静默无意义跑过。monkeypatch 重写会让"为啥要 max"这个 design intent 在测试里消失，跳过。

`feedback_pr_reviewer_independence` 原则：评审建议默认独立判断，不盲从。这两条都是风格 tradeoff 而非正确性 bug，hold 原状。

File: tests/unit/test_ai_aware_stage1_path_b.py [hunk 20]
```diff
@@ -0,0 +1,1541 @@
...
+            'text': '主人喜欢猫', 'importance': 7, 'entity': 'master',
+            'source': 'ai_disclosure',
+        },
+        {
+            'text': '主人喜欢猫', 'importance': 7, 'entity': 'master',
+            'source': 'user_observation',
+        },
+    ]
+
+    out = await fs._apersist_new_facts(
```
[the remark this answers is not in the document]
Comment (wehos) [reply tests/unit/test_ai_aware_stage1_path_b.py]: 已修 (commit `2f741bd2c`)：删掉 `out =` 绑定，直接 `await`。测试断言看的是 `fs._facts` 副作用不看返回值。

File: app/memory_server.py [hunk 21]
```diff
@@ -1760,6 +1844,238 @@ async def _adispatch_evidence_signals(
...
+        # AI-only msg（Codex P2 round-6 on PR #1408）。
+        # 修法：取 max(piggyback 节律, turn-count 节律) × IDLE_MIN 当估算
+        # 上限。默认下 max(3, 10) × 10min = 100min。LIMIT 兜底防爆 prompt，
+        # Stage-1 dedup hash 防双写——overshoot 是安全的。
+        cold_start_ticks_estimate = max(
+            EVIDENCE_AI_AWARE_EVERY_N_A_TICKS,
+            EVIDENCE_SIGNAL_CHECK_EVERY_N_TURNS,
+        )
+        estimated_a_coverage = timedelta(
+            minutes=cold_start_ticks_estimate * EVIDENCE_SIGNAL_CHECK_IDLE_MINUTES
```
[the remark this answers is not in the document]
Comment (wehos) [reply app/memory_server.py:1917]: 这轮不动，hold 在 round-6 的 `max(N_TICKS, N_TURNS) × IDLE_MIN` 形态。理由：

**Bug 是真的，但边界被低估**。具体触发链：长 uptime + sparse turn（user 几小时一次） + b_tick_counter 一直没达 N → A 已经多 tick 各自处理多小时窗口、B 一直没 trigger。第一次 B trigger 时 cold-start 100min 估算落在 A 真处理范围内，那段前的 AI-only msg 永久 skip。

**为什么不在本 PR 修**：

1. **proper fix 需要在 state 里追 A 的"processed-range start"（不是 end）**。`last_check_ts` 只记 A 上次 mark_done 的时间点 = A 窗口上界；A 窗口的下界（previous `last_check_ts`）在 mark_done 时被覆盖，没保留。要修就得加 state 字段 + 在 `_signal_check_mark_done` 之前写一道 snapshot。非平凡的 state plumbing。

2. **cap-and-cope（再加个绝对 floor 如 24 小时）违背用户偏好**。round-6 我们刚明确"derive from existing constants, no new magic LOOKBACK"，CodeRabbit 还专门为此存了 learning。再换个魔法常量是回头路。

3. **打包到 spawned liveness follow-up**（github.com 已 spawn 用户提的"毒窗口 max-retry 兜底"task）。那个 task 里已经攒了：
   - Codex round-4 P2: 同 ts 簇 LIMIT 截断丢 tail
   - path A / rebuttal loop 的 empty-rows transient 失败误判（round-5 仅修了 path B）
   - 现在这条: cold-start 不锚 A 真实 processed range
   - 用户的: 毒窗口无限重试
   
   都是"窗口覆盖 / coverage edge case"同族议题，散修 PR 越改越歪，应当统一设计 state-tracking + max-retry + bypass 策略一起实施。

**当前 PR scope hold**：path B 设计本身 + round-1..10 个 fix 已经覆盖了能就地修的所有 edge case。
